ocaml/typing/jkind.ml

```diff
@@ -322,9 +377,30 @@ module Const = struct
 
     let any = { jkind = not_mode_crossing Any; name = "any" }
 
-    let value = { jkind = not_mode_crossing Layout.Const.value; name = "value" }
+    let any_non_null =
+      let jkind =
+        { (not_mode_crossing Any) with nullability_upper_bound = Non_null }
+      in
+      { jkind; name = "any_non_null" }
+
+    let value_or_null =
+      { jkind = not_mode_crossing Layout.Const.value; name = "value_or_null" }
 
-    let void = { jkind = not_mode_crossing Layout.Const.void; name = "void" }
+    let value =
+      let jkind =
+        { (not_mode_crossing Layout.Const.value) with
+          nullability_upper_bound = Non_null
+        }
+      in
+      { jkind; name = "value" }
+
+    let void =
+      let jkind =
+        { (not_mode_crossing Layout.Const.void) with
+          nullability_upper_bound = Non_null
```

Comment:
In context of this PR, `Non_null` means "can't be a null pointer", and `Or_null` means "can be a null pointer". Under our design constraints, array type arguments must be non-null. Your interpretation would break that constraint. This does not apply specifically to `void`, but I decided to use `Non_null` here for consistency.

We don't have a design for `'a or_null` on non-value layouts, and the `or_null` type I'm adding in subsequent PRs is restricted to values. I think we can decide this question later.